File: Sagar_Dataharvesting/IEOD/IEOD V2/broker.py
import os
from datetime import datetime
import json
# import socketio
from io import StringIO
import requests
import pandas as pd
# from async_socket import MDSocket_io
class XTSlogin:
    def __init__(self):
        self.market_url = "https://ttblaze.iifl.com/apimarketdata/auth/login" 
        self.master_instruments_url = "https://ttblaze.iifl.com/apimarketdata/instruments/master"
        self.interactive_url = "https://ttblaze.iifl.com/interactive/user/session" 
        self.index_list_url = "https://ttblaze.iifl.com/marketdata/instruments/indexlist"
        self.quote_url = "https://ttblaze.iifl.com/marketdata/instruments/quotes"
        self.cover_order_url = "https://ttblaze.iifl.com/interactive/orders/cover"
        self.limit_order_url = "https://ttblaze.iifl.com/interactive/orders"
        self.historical_url =  "https://ttblaze.iifl.com/apimarketdata/instruments/ohlc"
        self.headers = {
            "Content-Type": "application/json"
        }
    def market_login(self, secret_key, app_key):
        payload = {
            "secretKey": secret_key,
            "appKey": app_key,
            "source": "WebAPI"
        }
        print(payload)

        try:
            response = requests.post(self.market_url, json=payload, headers=self.headers,verify=False)
            data = response.json()
            if data['type'] == 'success':
                token = data['result']['token']
                userid = data['result']['userID']
                self.market_token = token
                return token, userid
        except Exception as e:
            print("Error:", e)
            return None
    def interactive_login(self, secret_key, app_key):
        payload = {
            "secretKey": secret_key,
            "appKey": app_key,
            "source": "WebAPI"
        }

        try:
            response = requests.post(self.interactive_url, json=payload, headers=self.headers)
            data = response.json()
            if data['type'] == 'success':
                token = data['result']['token']
                userid = data['result']['userID']
                self.interactive_token = token
                return token, userid
        except Exception as e:
            print("Error:", e)
            return None
    def get_master(self, exchangeSegmentList):
        try:
            # payload = exchangeSegmentList

            response = requests.post(self.master_instruments_url, json = exchangeSegmentList)

            instrument_details = response.json()['result']
            # print(instrument_details)
            if exchangeSegmentList['exchangeSegmentList']==['NSECM']:
                iifl_instruments = pd.read_csv(StringIO(instrument_details), sep = '|', usecols=range(19), header = None, low_memory=False)
                # new_column_names = ["ExchangeSegment","ExchangeInstrumentID","InstrumentType","Name","Description","Series","NameWithSeries","InstrumentID","PriceBand.High","PriceBand.Low","FreezeQty","TickSize","LotSize","Multiplier","UnderlyingInstrumentId","UnderlyingIndexName","ContractExpiration","StrikePrice","OptionType","DisplayName","PriceNumerator"]
                new_column_names = [
                    "ExchangeSegment",
                    "ExchangeInstrumentID",
                    "InstrumentType",
                    "Name",
                    "Description",
                    "Series",
                    "NameWithSeries",
                    "InstrumentID",
                    "PriceBand.High",
                    "PriceBand.Low",
                    "FreezeQty",
                    "TickSize",
                    "LotSize",
                    "Multiplier",
                    "DisplayName",
                    "ISIN",
                    "PriceNumerator",
                    "PriceDenominator",
                    "DetailedDescription"
                ]

                iifl_instruments.columns = new_column_names
                iifl_instruments.to_csv('nsecm.csv')
            elif exchangeSegmentList['exchangeSegmentList']==['NSEFO']:
                iifl_instruments = pd.read_csv(StringIO(instrument_details), sep = '|', usecols=range(21), header = None, low_memory=False)
                new_column_names = ["ExchangeSegment","ExchangeInstrumentID","InstrumentType","Name","Description","Series","NameWithSeries","InstrumentID","PriceBand.High","PriceBand.Low","FreezeQty","TickSize","LotSize","Multiplier","UnderlyingInstrumentId","UnderlyingIndexName","ContractExpiration","StrikePrice","OptionType","DisplayName","PriceNumerator"]
                iifl_instruments.columns = new_column_names
                iifl_instruments['ContractExpiration'] = pd.to_datetime(iifl_instruments['ContractExpiration']).dt.date
                iifl_instruments.to_csv('nfo.csv')
            return iifl_instruments
        except Exception as e:
            print(e)

    def update_master_db(self):
        iifl_master = self.get_master({'exchangeSegmentList': ['NSEFO']})
        iifl_master = self.get_master({'exchangeSegmentList': ['NSECM']})
        # iifl = iifl_master[['ExchangeInstrumentID', 'Description', 'ContractExpiration','UnderlyingIndexName', 'StrikePrice', 'OptionType', 'ExchangeSegment', 'FreezeQty', 'LotSize', 'Series']]
        # iifl.to_csv('nfo.csv')
        print('master csv updated')
    def get_index_list(self, exchangeSegment):
        try:
            response = requests.get(self.index_list_url, params = {"exchangeSegment": exchangeSegment})
            data = response.json()
            if data['type']=="success":
                return data["result"]
            else:
                return None
        except Exception as e:
            print("Error:", e)
            return None
        
    def get_quotes(self, instruments):
            headers = self.headers
            headers.update({'Authorization': self.market_token})
            payload = {"instruments": instruments, "xtsMessageCode": 1512, "publishFormat" : "JSON" }
            try:
                response = requests.post(self.quote_url,json= payload, headers=headers)
                # print(response.json())
                return response.json()
            except Exception as e:
                print(e)
                return None
    def place_cover_order(self, params):
        headers = self.headers
        headers.update({'Authorization': self.interactive_token})
        # payload = params
        try:
            params["exchangeSegment"]="NSEFO"
            # params["productType"] ="MIS"
            params["orderType"]="LIMIT"
            # params["timeInForce"]="DAY"
            params["disclosedQuantity"]=0
            # params["limitPrice"]=0
            params["orderUniqueIdentifier"]= f'{params["exchangeInstrumentID"]}'
            
            payload = params
            response = requests.post(self.cover_order_url, json = payload,headers=headers)
            return response.json()
        except Exception as e:
            print(e)
            return None
        
    def place_limit_order(self, params):
        headers = self.headers
        headers.update({'Authorization': self.interactive_token})
        try:
            params["exchangeSegment"]="NSEFO"
            params["productType"] ="NRML"
            params["orderType"]="LIMIT"
            params["timeInForce"]="DAY"
            params["disclosedQuantity"]=0
            params["orderUniqueIdentifier"]= f'{params["exchangeInstrumentID"]}'
            payload = params
            response = requests.post(self.limit_order_url, json = payload,headers=headers)
            return response.json()
        except Exception as e:
            print(e)
            return None
        
    def subscribe_symbols(self, instruments,market_token):
        subscription_url = "https://ttblaze.iifl.com/apimarketdata/instruments/subscription"

        params = {'instruments': instruments, 'xtsMessageCode': 1512}
        session = requests.Session()
        headers = self.headers
        headers.update({'Authorization': self.market_token})
        
        try:
            r = session.request("POST",
                                        subscription_url,
                                        data=json.dumps(params),
                                        params=None,
                                        headers=headers
                                        )
            # print(r.content)
        except Exception as e:
            raise e
        
    def get_historical_data(self, params):
        headers = self.headers
        headers.update({'Authorization': self.market_token})
        try:
            params = params
            response = requests.get(self.historical_url, params=params, headers=headers)
            
            if response.status_code == 200:
                return response.json()
            else:
                print("Failed to retrieve data. Status code:", response.status_code)
                print("Response:", response.text)
                return None
        except Exception as e:
            print("Exception occurred:", e)
            return None
    def update_master_db(self):
        iifl_master = self.get_master({'exchangeSegmentList': ['NSEFO']})
        # iifl = iifl_master[['ExchangeInstrumentID', 'Description', 'ContractExpiration','UnderlyingIndexName', 'StrikePrice', 'OptionType', 'ExchangeSegment', 'FreezeQty', 'LotSize', 'Series']]
        # iifl.to_csv('nfo.csv')
        iifl_master = self.get_master({'exchangeSegmentList': ['NSECM']})

File: Sagar_Dataharvesting/Real-Time DB/data harvesting node/broker.js
const axios = require("axios");
const fs = require("fs");
const path = require("path");
const { parse } = require("csv-parse/sync");

class XTS {
  constructor() {
    this.marketUrl = "https://ttblaze.iifl.com/apimarketdata/auth/login";
    this.masterInstrumentsUrl = "https://ttblaze.iifl.com/apimarketdata/instruments/master";
    this.subscriptionUrl = "https://ttblaze.iifl.com/apimarketdata/instruments/subscription"
    this.headers = { "Content-Type": "application/json" };
    this.xtsSubscribedSymbols = [];
    this.tokenFilePath = path.join(__dirname, "xts_tokens.json");
    this.nfoCsvPath = path.join(__dirname, "nfo.csv");

    const tokens = this.loadTokensFromFile();
    if (tokens && tokens.date === new Date().toISOString().split("T")[0]) {
      this.marketToken = tokens.marketToken;
      this.interactiveToken = tokens.interactiveToken;
      this.userid = tokens.userID;
      console.log("Tokens loaded from xts_tokens.json");
    }
  }

  // Save tokens to a JSON file
  saveTokensToFile(tokens) {
    try {
      fs.writeFileSync(this.tokenFilePath, JSON.stringify(tokens, null, 2), "utf-8");
      console.log("Tokens saved to xts_tokens.json");
    } catch (error) {
      console.error("Failed to save tokens:", error.message);
    }
  }

  // Load tokens from JSON file
  loadTokensFromFile() {
    try {
      if (fs.existsSync(this.tokenFilePath)) {
        const tokens = JSON.parse(fs.readFileSync(this.tokenFilePath, "utf-8"));
        console.log("Tokens loaded from xts_tokens.json");
        return tokens;
      }
    } catch (error) {
      console.error("Failed to load tokens:", error.message);
    }
    return null;
  }
  async getQuotes(instruments) {
    const headers = { ...this.headers, Authorization: this.marketToken };
    const payload = {
      instruments: instruments,
      xtsMessageCode: 1512,
      publishFormat: "JSON",
    };
  
    try {
      const response = await axios.post(
        "https://ttblaze.iifl.com/marketdata/instruments/quotes",
        payload,
        { headers }
      );
  
      // Log and return the response data
      console.log("Quotes Response Data:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching quotes:", error.message);
      if (error.response) {
        // Log additional error details if available
        console.error("Error Response Data:", error.response.data);
        console.error("Error Response Status:", error.response.status);
        console.error("Error Response Headers:", error.response.headers);
      }
      return null;
    }
  }
  
  // Market login
  async marketLogin(secretKey, appKey) {
    try {
      const response = await axios.post(this.marketUrl, {
        secretKey,
        appKey,
        source: "WebAPI",
      });
      const data = response.data;
      if (data.type === "success") {
        this.marketToken = data.result.token;
        this.userid = data.result.userID;
        console.log("Market login successful");
        this.saveTokensToFile({
          marketToken: this.marketToken,
          interactiveToken: this.interactiveToken,
          userID: this.userid,
          date: new Date().toISOString().split("T")[0],
        });
        return { token: this.marketToken, userID: this.userid };
      } else {
        console.error("Market login failed:", data.message);
        return null;
      }
    } catch (error) {
      console.error("Market login error:", error.response?.data || error.message);
      return null;
    }
  }

  // Fetch master instruments and save to nfo.csv
  

  // Parse instrument data using csv-parse/sync
  
  async subscribeSymbols(instruments) {
    try {
      const newInstruments = instruments.filter(
        (instrument) => !this.xtsSubscribedSymbols.includes(instrument)
      );
      if (newInstruments.length > 0) {
        const response = await axios.post(
          this.subscriptionUrl,
          {
            instruments: newInstruments,
            xtsMessageCode: 1512,
          },
          { headers: { Authorization: this.marketToken } }
        );
  
        // Log the entire response
        // console.log("Response:", response);
  
        // Log specific parts of the response
        console.log("Response Data:", response.data);
        // console.log("Response Status:", response.status);
        // console.log("Response Headers:", response.headers);
  
        this.xtsSubscribedSymbols.push(...newInstruments);
        // console.log(this.xtsSubscribedSymbols);
        // console.log("Subscribed to new instruments.");
      } else {
        console.log("No new instruments to subscribe.");
      }
    } catch (error) {
      console.error("Failed to subscribe symbols:", error.message);
      if (error.response) {
        // Log error response details if available
        console.error("Error Response Data:", error.response.data);
        console.error("Error Response Status:", error.response.status);
        console.error("Error Response Headers:", error.response.headers);
      }
    }
  }
  
  // Save parsed data to a CSV file
  saveToCsv(data, filePath) {
    try {
      const headers = Object.keys(data[0]).join(",");
      const csvContent = [headers, ...data.map((row) => Object.values(row).join(","))].join("\n");
      fs.writeFileSync(filePath, csvContent, "utf-8");
      console.log(`Data successfully saved to ${filePath}`);
    } catch (error) {
      console.error("Error saving data to CSV:", error.message);
    }
  }
}

module.exports = XTS;




